### PR TITLE
[MNG-5726] Support regular expression matching in profile activation for

### PR DESCRIPTION
**Pull request by kwin**
OS version

This requires using the reserved prefix "regex:" in the version element.

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[MNG-XXX] SUMMARY`,
       where you replace `MNG-XXX` and `SUMMARY` with the appropriate JIRA issue.
 - [ ] Also format the first line of the commit message like `[MNG-XXX] SUMMARY`.
       Best practice is to use the JIRA issue title in both the pull request title and in the first line of the commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by gnodet** [comment]
Would it make sense to also support the easier `glob:` syntax then ?

**Comment by kwin** [comment]
@gnodet Is there general support for matching globs built into Java (apart from Files NIO)? I haven't found anything and quick googling revealed https://stackoverflow.com/questions/1247772/is-there-an-equivalent-of-java-util-regex-for-glob-type-patterns. So I would say, regex is good enough for now.

**Comment by gnodet** [comment]
> @gnodet Is there general support for matching globs built into Java (apart from Files NIO)? I haven't found anything and quick googling revealed https://stackoverflow.com/questions/1247772/is-there-an-equivalent-of-java-util-regex-for-glob-type-patterns. So I would say, regex is good enough for now.
> 
> (I don't want to rely on external libraries like plexus-utils or maven-utils for such stuff TBH)

Fwiw, a workaround is already used in maven: https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/ExclusionArtifactFilter.java
Maybe we can move the code in its own class in maven-core and reuse it ?

**Comment by kwin** [comment]
@gnodet Interesting. We can support `glob:` later once that refactoring is done. For now I would go with just `regex:`. Any concerns with that?

**Comment by gnodet** [comment]
> @gnodet Interesting. We can support `glob:` later once that refactoring is done. For now I would go with just `regex:`. Any concerns with that?

None...

**Review by gnodet — APPROVED** [review]

File: maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/OperatingSystemProfileActivator.java [hunk 1]
```diff
@@ -53,17 +57,21 @@ public boolean isActive(Profile profile, ProfileActivationContext context, Model
 
         boolean active = ensureAtLeastOneNonNull(os);
 
+        String actualOsName = context.getSystemProperties().get("os.name").toLowerCase(Locale.ENGLISH);
```
Comment (gnodet) [inline maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/OperatingSystemProfileActivator.java:60]: Why not `Os.OS_NAME` and related ?

Comment (kwin) [reply maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/OperatingSystemProfileActivator.java:60]: To streamline the testing approach with the other activators. `Os.OS_NAME` cannot be easily overwritten.

Comment (kwin) [reply maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/OperatingSystemProfileActivator.java:60]: Compare with https://github.com/apache/maven/blob/ab6ec5bd74af20ab429509eb56fc8e3dff4c7fc7/maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/JdkVersionProfileActivator.java#L64 and https://github.com/apache/maven/blob/ab6ec5bd74af20ab429509eb56fc8e3dff4c7fc7/maven-model-builder/src/test/java/org/apache/maven/model/profile/activation/JdkVersionProfileActivatorTest.java#L50

**Comment by jira-importer** [comment]
Resolve #6587

**Pull request merged**